pkg/services/ngalert/store/range_to_instant.go

```diff
@@ -13,34 +14,48 @@ type dsType struct {
 	DS struct {
 		Type string `json:"type"`
 	} `json:"datasource"`
+	Range bool `json:"range"`
 }
 
 func (t dsType) isLoki() bool {
 	return t.DS.Type == datasources.DS_LOKI
 }
 
-// canBeInstant checks if any of the query nodes that are loki range queries can be migrated to instant queries.
+func (t dsType) isPrometheus() bool {
+	return t.DS.Type == datasources.DS_PROMETHEUS
+}
+
+type optimization struct {
+	// Index of the query that can be optimized
+	i int
+	// Type of the query that ca be optimized (loki,prometheus)
+	t string
+}
+
+// canBeInstant checks if any of the query nodes that are loki or prometheus range queries can be migrated to instant queries.
 // If any are migratable, those indices are returned.
-func canBeInstant(r *models.AlertRule) ([]int, bool) {
+func canBeInstant(r *models.AlertRule) ([]optimization, bool) {
 	if len(r.Data) < 2 {
 		return nil, false
 	}
 	var (
-		optimizableIndices []int
+		optimizableIndices []optimization
 		canBeOptimized     = false
 	)
-	// Loop over query nodes to find all Loki range queries.
+	// Loop over query nodes to find all range queries.
 	for i := range r.Data {
-		if r.Data[i].QueryType != "range" {
-			continue
-		}
 		var t dsType
 		// We can ignore the error here, the query just won't be optimized.
 		_ = json.Unmarshal(r.Data[i].Model, &t)
 
-		if !t.isLoki() {
+		if !t.isLoki() && !t.isPrometheus() {
 			continue
 		}
+
+		if (t.isLoki() && r.Data[i].QueryType != "range") || (t.isPrometheus() && !t.Range) {
+			continue
+		}
```

Comment:
[nit] Stylistic comment - I find it a little jarring that we use `switch` below, but these `isXXX` functions here. Might look nicer to use a `switch` here too.

Reply:
SOLD